framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -1680,12 +1680,42 @@ public Void visitExecutable(AnnotatedExecutableType type, AnnotatedTypeFactory p
 
             // When visiting an executable type, skip the receiver so we
             // never inherit class annotations there.
-
-            // Also skip constructor return types (which somewhat act like
-            // the receiver).
             MethodSymbol methodElt = (MethodSymbol) type.getElement();
             if (methodElt == null || !methodElt.isConstructor()) {
                 scan(type.getReturnType(), p);
+            } else if (methodElt.isConstructor()) {
+                // If a constructor declaration is not explicitly annotated,
+                // annotate it with the type on the class declaration.
+                AnnotatedDeclaredType returnType = (AnnotatedDeclaredType) type.getReturnType();
+                // At this point defaults have been applied to class declarations but not
+                // to constructor return types. The check "returnType.getAnnotations().isEmpty()"
+                // returns true for constructors whose return types haven't been explicitly
+                // annotated.
+                // NOTE: change to getExplicitAnnotations() when http://tinyurl.com/cfissue/2324 is
```

Comment:
Could you abstract this out into a method?  Then the call can be changed to `getExplicitAnnotations()` in the future, and the hack will be better-encapsulated.